opal/mca/installdirs/runtime/opal_installdirs_runtime.c

```diff
@@ -0,0 +1,65 @@
+/*
+ * Copyright (c) 2006-2007 Los Alamos National Security, LLC.  All rights
+ *                         reserved.
+ * Copyright (c) 2007      Cisco Systems, Inc.  All rights reserved.
+ * $COPYRIGHT$
+ *
+ * Additional copyrights may follow
+ *
+ * $HEADER$
+ */
+
+#include "opal_config.h"
+
+#include <stdlib.h>
+#include <string.h>
+
+#include "opal/constants.h"
+#include "opal/mca/installdirs/installdirs.h"
+#include "opal/runtime/opal.h"
+
+static int installdirs_runtime_open(void);
+
+opal_installdirs_base_component_t mca_installdirs_runtime_component = {
+    /* First, the mca_component_t struct containing meta information
+       about the component itself */
+    {OPAL_INSTALLDIRS_BASE_VERSION_2_0_0,
+
+     /* Component name and version */
+     "runtime", OPAL_MAJOR_VERSION, OPAL_MINOR_VERSION, OPAL_RELEASE_VERSION,
+
+     /* Component open and close functions */
+     installdirs_runtime_open, NULL},
+    {/* This component is checkpointable */
+     MCA_BASE_METADATA_PARAM_CHECKPOINT},
+
+    /* Next the opal_install_dirs_t install_dirs_data information */
+    {
+        NULL,
+    },
+};
+
+#include <dlfcn.h>
+#include "opal/util/basename.h"
+
+static int installdirs_runtime_open(void)
+{
+    Dl_info info;
+    void* opal_fct;
+
+    /* Casting from void* to fct pointer according to POSIX.1-2001 and POSIX.1-2008 */
+    *(void **)&opal_fct = dlsym(RTLD_DEFAULT, "opal_init_util");
```

Comment:
What does this return if you have compiled statically? (i.e., libopen_pal.a)

Reply:
the path of the application, which is useless. That's why I only enabled this feature on shared buids.

Reply:
What happens if you configure with `--enable-static --enable-shared` (i.e., build both static *and* shared libraries)?

Reply:
You can do that ?

Reply:
Yes, this is a valid way to build.  For example:

```
$ ls $HOME/bogus2/lib/libmpi.*
/Users/jsquyres/bogus2/lib/libmpi.0.dylib*
/Users/jsquyres/bogus2/lib/libmpi.a
/Users/jsquyres/bogus2/lib/libmpi.dylib@
/Users/jsquyres/bogus2/lib/libmpi.la*
```

Are you sure that we're checking static vs. dynamic builds elsewhere in the codebase for exclusive behavior?  (i.e., assuming that we're only building static *or* dynamic?  I thought we took pains *not* to do that...?

Reply:
This means that all components that use `$enable_shared` to build themselves are doing it incorrectly. Bummer.

Reply:
I just confirmed this does not work at all for static builds. Static build can only be moved around if OPAL_PREFIX is correctly defined.

Reply:
> This means that all components that use $enable_shared to build themselves are doing it incorrectly. Bummer.

Can you list which components you're referring to?

> I just confirmed this does not work at all for static builds. Static build can only be moved around if OPAL_PREFIX is correctly defined.

I thought you said above that `dladdr()` would return the path of the application for static builds.  Is that not the case?

Reply:
yes, it returns the path of the application, but that's useless for us.